packages/interactor/test/create-interactor.test.ts

```diff
@@ -42,7 +45,22 @@ const TextField = createInteractor<HTMLInputElement>('text field')({
     value: (element) => element.value
   },
   actions: {
-    fillIn: (element, value: string) => { element.value = value }
+    fillIn: ({ element }, value: string) => { element.value = value },
+    click: ({ element }) => { element.click() }
+  }
+});
+
+const Datepicker = createInteractor<HTMLDivElement>("datepicker")({
+  selector: "div.datepicker",
+  defaultLocator: element => element.querySelector("label")?.textContent || "",
+  filters: {
+    open: element => !!element.querySelector("div.calendar"),
+    month: element => element.querySelector("div.calendar h4")?.textContent
+  },
+  actions: {
+    toggle: async ({ interactor }) => {
+      await interactor.find(TextField.byPlaceholder("YYYY-MM-DD")).click();
```

Comment:
Are actions allowed to be async?

Reply:
Yes, we would need to intentionally swallow the promise to not support it.

Reply:
Yeah, it's just a little weird because the idea originally that all element actions need to be synchronous so that the element reference does not get stale.

It feels like we should be using _either_ element _or_ interactor, but never both. Imagine you had something like:

```ts
actions: {
  toggle: async ({ interactor, element }) => {
    await interactor.find(TextField.byPlaceholder("YYYY-MM-DD")).click();
    element.data.clicked = true;
  }
}
```

`element` would be stale, and this should be an error, and in the case where you _do_ use element, a promise return value is unacceptable. What do you think?

Reply:
> element would be stale, and this should be an error, and in the case where you do use element, a promise return value is unacceptable.

This is starting to get a little complex for my liking. I don't think we should treat the usage of arguments differently based on what is being returned. That does not seem like it would make for a simple API.

Need to think about this more. Maybe we can talk tomorrow about it.

Reply:
Here's an idea:

1. Add an async method to the interactor interface that fetches the element; basically `unsafeSyncResolve()` wrapped in `converge()`
1. Provide only the interactor interface to actions
1. Don't run actions within a `converge()` (otherwise we would be doubling up)

A nice side-effect of this IMO is that actions would not be run more than once, which in my experience people have a hard time understanding/retaining.

Reply:
What would an example of this look like?

Reply:
Using your example problematic code from above:

```ts
actions: {
  toggle: async interactor => {
    await interactor.find(TextField.byPlaceholder("YYYY-MM-DD")).click();
    let element = await interactor.element();
    element.data.clicked = true;
  }
}
```

However, one could still introduce a bug like this:

```ts
actions: {
  toggle: async interactor => {
    let element = await interactor.element();
    await interactor.find(TextField.byPlaceholder("YYYY-MM-DD")).click();
    element.data.clicked = true;
  }
}
```

We could avoid this footgun with a callback API, but not sure it is desirable to mix callbacks and promises:

```ts
actions: {
  toggle: async interactor => {
    await interactor.find(TextField.byPlaceholder("YYYY-MM-DD")).click();
    await interactor.withElement(element => {
      // We would enforce this as synchronous
      element.data.clicked = true;
    });
  }
}
```

Reply:
of those options, I think that number 3 is the safest. where you have an interactor.perform() or interactor.withElement() method that lets you access the element in the context of full validation. 

One possible syntax is we treat actions as first class objects, and then if we see a function, we treat it as the base case of lifting a simple synchronous function into an action context:

```ts
import { action } from '@bigtest/interactor';

actions: {
  toggle: action(async (interactor) => {
    await interactor.find(TextField.byPlaceholder("YYYY-MM-DD")).click();
    await interactor.perform(element => element.data.clicked = true);
  })
}
```

but still allow for simple actions:

```ts
actions: {
  click: element => element.click()
}
```

which is just sugar for:

```ts
actions: {
  click: action(async (interactor) => await interactor.perform(element => element.click()))
}
```

This is easy to type at both runtime and compile time because the action would be either a `Function`, or an instance of `Action`:

```ts
type ActionSpec = Action | (element: Element) => void;
```

Reply:
I like this, but it feels like the inverse of what I would expect coming at this without all this context, and I think we can simplify it:

```ts
actions: {
  toggle: async interactor => {
    await interactor.find(TextField.byPlaceholder("YYYY-MM-DD")).click();
    await interactor.perform(element => element.data.clicked = true);
  }
}
```

and

```ts
actions: {
  toggle: perform(element => element.click())
}
```

where the sugar happens here:

```ts
function perform(fn) {
  return interactor => interactor.perform(fn);
}
```

Reply:
I ran with this to see what it would look like, no obligations. Typing was surprisingly easy.

Reply:
The default should align with the 80/20 or 90/10 case. The only problem is that we don't have any data. What do we think will be the most common usage?

Reply:
In this scenario I see `perform()` as a power user shortcut for low-level operations where we don't have a base interactor defined already. I'm hoping eventually we'll have base interactors and you won't need to `perform` very often.